### ISSUE TITLE
install_ffmpeg.sh fails with `ERROR: gnutls not found using pkg-config` on Ubuntu 20.04 LTS

### ISSUE THREAD

**Issue by chrishobcroft** [livepeer/go-livepeer#1500]
**Describe the bug**

When running `./install_ffmpeg.sh`, the process fails, reporting an error `ERROR: gnutls not found using pkg-config`

**To Reproduce**

1. Run through this: https://github.com/livepeer/go-livepeer/blob/master/doc/install.md#option-2-build-from-source

2. At step 4, when running `./install_ffmpeg.sh` it fails with `ERROR: gnutls not found using pkg-config`

**Expected behavior**

I expect that `./install_ffmpeg.sh` should successfully install `FFmpeg`.

I also expect that the doc https://github.com/livepeer/go-livepeer/blob/master/doc/install.md should provide instructions to help open-source software developers to build Livepeer from source, and that these instructions should work in reasonable circumstances.

I also expect there to be an `./uninstall_ffmpeg.sh` because a) installers fail, as above, b) installers sometimes install software which messes with the operating system, and you don't want to have to reinstall your operating system :|

Finally, and this is definitely an aside, is to expect that this should work with vanilla `ffmpeg`, instead of requiring some home-baked version. But I can't remember why we have to roll our own - perhaps someone can remind me?

**Screenshots**

![Screenshot from 2020-05-17 18-31-41](https://user-images.githubusercontent.com/2212651/82147340-adeb7280-986c-11ea-89cc-7170e64df738.png)

**Desktop (please complete the following information):**

![image](https://user-images.githubusercontent.com/2212651/82147623-3ec24e00-986d-11ea-8660-8b0782d45a2d.png)

**Additional context**
Here is full output of session:
```
ubuntu@cibulka:~/livepeer/go-livepeer$ ./install_ffmpeg.sh
++ uname
+ [[ Linux == *\M\S\Y\S* ]]
+ export PATH=/home/ubuntu/compiled/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
+ PATH=/home/ubuntu/compiled/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
+ export PKG_CONFIG_PATH=:/home/ubuntu/compiled/lib/pkgconfig
+ PKG_CONFIG_PATH=:/home/ubuntu/compiled/lib/pkgconfig
++ uname
+ '[' Linux '!=' Darwin ']'
+ '[' '!' -e /home/ubuntu/nv-codec-headers ']'
+ '[' '!' -e /home/ubuntu/nasm/nasm ']'
+ '[' '!' -e /home/ubuntu/x264 ']'
++ uname
+ [[ Linux != *\M\S\Y\S* ]]
+ '[' '!' -e /home/ubuntu/gmp-6.1.2 ']'
+ '[' '!' -e /home/ubuntu/nettle-3.4.1 ']'
+ '[' '!' -e /home/ubuntu/gnutls-3.5.18 ']'
+ EXTRA_FFMPEG_FLAGS=
+ EXTRA_LDFLAGS=
++ uname
+ '[' Linux == Darwin ']'
+ which clang
+ '[' '!' -e /home/ubuntu/ffmpeg/libavcodec/libavcodec.a ']'
+ git clone https://git.ffmpeg.org/ffmpeg.git /home/ubuntu/ffmpeg
fatal: destination path '/home/ubuntu/ffmpeg' already exists and is not an empty directory.
+ echo 'FFmpeg dir already exists'
FFmpeg dir already exists
+ cd /home/ubuntu/ffmpeg
+ git checkout 3ea705767720033754e8d85566460390191ae27d
HEAD is now at 3ea7057677 nvenc: implement flush to help allow an encoder to be re-used
+ ./configure --fatal-warnings --disable-programs --disable-doc --disable-sdl2 --disable-iconv --disable-muxers --disable-demuxers --disable-parsers --disable-protocols --disable-encoders --disable-decoders --disable-filters --disable-bsfs --disable-postproc --disable-lzma --enable-gnutls --enable-libx264 --enable-gpl --enable-protocol=https,rtmp,file --enable-muxer=mpegts,hls,segment,mp4 --enable-demuxer=flv,mpegts,mp4,mov --enable-bsf=h264_mp4toannexb,aac_adtstoasc,h264_metadata,h264_redundant_pps --enable-parser=aac,aac_latm,h264 --enable-filter=abuffer,buffer,abuffersink,buffersink,afifo,fifo,aformat --enable-filter=aresample,asetnsamples,fps,scale --enable-encoder=aac,libx264 --enable-decoder=aac,h264 --extra-cflags=-I/home/ubuntu/compiled/include '--extra-ldflags=-L/home/ubuntu/compiled/lib ' --prefix=/home/ubuntu/compiled
ERROR: gnutls not found using pkg-config

If you think configure made a mistake, make sure you are using the latest
version from Git.  If the latest version fails, report the problem to the
ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.freenode.net.
Include the log file "ffbuild/config.log" produced by configure as this will help
solve the problem.
ubuntu@cibulka:~/livepeer/go-livepeer$ 
```

**Comment by j0sh** [comment]
> I also expect there to be an ./uninstall_ffmpeg.sh because a) installers fail, as above, b) installers sometimes install software which messes with the operating system,

Note that our installer goes to lengths *not* to mess up your operating system; that's why everything is downloaded, built and installed $HOME. This is the same reason why we don't provide an uninstall script; it's somewhat common for folks to have, say a `$HOME/ffmpeg` from other projects, and we don't want users to make the mistake of nuking that.

That being said, we could probably improve things by doing something like placing builds elsewhere, eg within the `go-livepeer` directory, rather than directly within $HOME. We would have to make sure that doesn't break CI caching though. (This script was originally intended for CI, rather than human users.)

For the time being, try doing either of the following:
* `rm -rf` the directories that the build script is skipping over in the `[ ! -e ...` lines . Be sure you're not actually using anything in there.
* Replace the $HOME env var and re-run the script so that the build output gets placed there, eg `$HOME=$HOME/Desktop/livepeer-dependencies`  . Be sure the bindir exists in PATH as well, eg $PATH=$HOME/compiled/bin:$PATH`

> is to expect that this should work with vanilla ffmpeg, instead of requiring some home-baked version. But I can't remember why we have to roll our own - perhaps someone can remind me?


We work off a version of ffmpeg that's pinned from a particular commit in their tree. The only "rolling" we're doing is compiling our own builds. There are a few reasons for this:
* Package distros lag a lot; we usually need more recent changes, especially things that haven't made it into a release yet.
* ffmpeg's API / behavior isn't the most stable, so pinning that allows us to ensure builds are somewhat reproducible and don't randomly break.
* ffmpeg by default incorporates a lot of unrelated functionality, and our build trims that down to what's necessary for livepeer. This keeps binary sizes manageable, because we ship static builds without expecting the user to have ffmpeg installed on their own system.

**Comment by criticaltv** [comment]
Thank you for the explanation.

I'm less interested right now on getting it working for myself. More interested in getting it to work for someone coming to this project for the first time.

But wrt your suggestion "For the time being, try doing either of the following:" the first option seems incomplete? Or do I do both not either?

**Comment by chrishobcroft** [comment]
Thank you for the explanation.

I'm less interested right now on getting it working for myself (although that would be a good start and could spark a career change for me, and I might even try to prototype this: https://github.com/livepeer/go-livepeer/issues/1504!)

I am more interested in getting it to work for someone with far more skills in software than I have, coming to this project for the first time, and wanting to extend it.

My objective would be to make it so that they could conceivably start to make contributions to this open-source project.

As it stands, if the process doesn't work, then the barrier-to-entry is really quite high...

But wrt your suggestion "For the time being, try doing either of the following:" the first option seems incomplete? Or do I do "both" (not "either")?

**Comment by chrishobcroft** [comment]
I did the `rm -rf` option, and then ran the `./install_ffmpeg.sh` again, and this time, different error:

```
ubuntu@cibulka:~/livepeer/go-livepeer$ ./install_ffmpeg.sh
++ uname
+ [[ Linux == *\M\S\Y\S* ]]
+ export PATH=/home/ubuntu/compiled/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/go/bin
+ PATH=/home/ubuntu/compiled/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/go/bin
+ export PKG_CONFIG_PATH=:/home/ubuntu/compiled/lib/pkgconfig
+ PKG_CONFIG_PATH=:/home/ubuntu/compiled/lib/pkgconfig
++ uname
+ '[' Linux '!=' Darwin ']'
+ '[' '!' -e /home/ubuntu/nv-codec-headers ']'
+ git clone https://git.videolan.org/git/ffmpeg/nv-codec-headers.git /home/ubuntu/nv-codec-headers
Cloning into '/home/ubuntu/nv-codec-headers'...
remote: Enumerating objects: 526, done.
remote: Counting objects: 100% (526/526), done.
remote: Compressing objects: 100% (452/452), done.
remote: Total 526 (delta 300), reused 0 (delta 0)
Receiving objects: 100% (526/526), 115.27 KiB | 318.00 KiB/s, done.
Resolving deltas: 100% (300/300), done.
+ cd /home/ubuntu/nv-codec-headers
+ git checkout 250292dd20af60edc6e0d07f1d6e489a2f8e1c44
Note: switching to '250292dd20af60edc6e0d07f1d6e489a2f8e1c44'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 250292d Bump for (in-dev) 9.1.23.2
+ make -e PREFIX=/home/ubuntu/compiled
sed 's#@@PREFIX@@#/home/ubuntu/compiled#' ffnvcodec.pc.in > ffnvcodec.pc
+ make install -e PREFIX=/home/ubuntu/compiled
sed 's#@@PREFIX@@#/home/ubuntu/compiled#' ffnvcodec.pc.in > ffnvcodec.pc
install -m 0755 -d '/home/ubuntu/compiled/include/ffnvcodec'
install -m 0644 include/ffnvcodec/*.h '/home/ubuntu/compiled/include/ffnvcodec'
install -m 0755 -d '/home/ubuntu/compiled/lib/pkgconfig'
install -m 0644 ffnvcodec.pc '/home/ubuntu/compiled/lib/pkgconfig'
+ '[' '!' -e /home/ubuntu/nasm/nasm ']'
+ git clone -b nasm-2.14.02 https://repo.or.cz/nasm.git /home/ubuntu/nasm
fatal: destination path '/home/ubuntu/nasm' already exists and is not an empty directory.
```

**Comment by j0sh** [comment]
> fatal: destination path '/home/ubuntu/nasm' already exists and is not an empty directory.

Was `/home/ubuntu/nasm` also rm -rf'd ? Sounds like it was still there, so the clone failed.

**Comment by chrishobcroft** [comment]
OK, I deleted all the folders in `~` created by ffmpeg_install.sh.

I then ran it again, and it created the same folders over again, and failed with `ERROR: gnutls not found using pkg-config` as per the original issue report.

**Comment by chrishobcroft** [comment]
OK, update on this - I just tried again, starting from a clean `Ubuntu 20.04`, and the `./install_ffmpeg.sh` still fails with `ERROR: gnutls not found using pkg-config`

These are the commands I ran:
```
sudo apt-get update
sudo apt-get -y install build-essential pkg-config autoconf gnutls-dev git curl
mkdir livepeer && cd livepeer
git clone https://github.com/livepeer/go-livepeer.git
cd go-livepeer
./install_ffmpeg.sh
```

Full session log attached. [session log - livepeer build.txt](https://github.com/livepeer/go-livepeer/files/4714800/session.log.-.livepeer.build.txt)

![image](https://user-images.githubusercontent.com/2212651/83480273-77426880-a4b8-11ea-9058-9c9abb19a53c.png)

It's not currently possible to follow the published instructions and build livepeer from source.

How can this be resolved?
